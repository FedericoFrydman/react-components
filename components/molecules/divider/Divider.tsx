import { Ref, forwardRef } from 'react';
import cn from 'classnames';

import { Text } from '../../atoms/text/Text';
import './divider.scss';

export type DividerTint = 'light' | 'medium' | 'dark';

export interface DividerProps extends Omit<React.HTMLProps<HTMLDivElement>, 'css'> {
  /**
   * Define the text that would be displayed in divider
   */
  label?: string;
  /**
   * Tint of the divider. Defaults to `light`.
   */
  tint?: DividerTint;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
}

const DividerComponent = (
  { label, tint = 'light', classNames = [], ...props }: DividerProps,
  ref: DividerProps['ref'],
) => {
  const classDivider = 'lsux-divider';
  const cnDivider = cn(
    {
      [`${classDivider}`]: true,
      [`${classDivider}__text`]: label,
      [`${classDivider}__line`]: !label,
      [`${classDivider}--tint-${tint}`]: tint,
    },
    classNames,
  );

  return (
    <div {...props} className={cnDivider} ref={ref}>
      <Text text={label} as="p" />
    </div>
  );
};

export const Divider = forwardRef<HTMLDivElement, DividerProps>(DividerComponent);
