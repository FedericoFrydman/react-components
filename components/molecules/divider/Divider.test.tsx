import { createRef } from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { render, screen } from '@testing-library/react';
import cn from 'classnames';

import { Divider, DividerProps } from './Divider';

describe('Divider Component', () => {
  const classDivider = 'lsux-divider';
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Divider label="or" />);
    testId = 'test-divider';
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('no tint is set', () => {
        const className = `${classDivider}--tint-light`;

        it('defaults tint to light', () => {
          const Divider = renderDivider();

          expect(Divider).toHaveClass(className);
        });
      });
      describe('tint is set', () => {
        const className = `${classDivider}--tint-medium`;

        it('sets the tint class to the selected tint', () => {
          const Divider = renderDivider({ tint: 'medium' });

          expect(Divider).toHaveClass(className);
        });
      });
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the Divider the provided classNames', () => {
          const Divider = renderDivider({ classNames });

          expect(Divider).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });

    describe('label', () => {
      describe('when label is provided', () => {
        const label = 'test';
        const dividerTextClass = `${classDivider}__text`;

        it('gives the Divider the text className', () => {
          const Divider = renderDivider({ label });

          expect(Divider).toHaveClass(dividerTextClass, { exact: false });
        });
      });

      describe('when label is not provided', () => {
        const label = '';
        const dividerLineClass = `${classDivider}__line`;

        it('gives the Divider the line className', () => {
          const Divider = renderDivider({ label });

          expect(Divider).toHaveClass(dividerLineClass, { exact: false });
        });
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLDivElement>();

    it('provides a ref to the divider', () => {
      expect(renderDivider({ ref })).toEqual(ref.current);
    });
  });

  const renderDivider = (props: DividerProps = {}): HTMLElement => {
    render(<Divider {...props} data-testid={testId} />);

    return screen.getByTestId(testId);
  };
});
