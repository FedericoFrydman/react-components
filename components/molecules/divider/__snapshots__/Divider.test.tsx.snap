// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider Component should render the component 1`] = `
<div
  className="lsux-divider lsux-divider__text lsux-divider--tint-light"
>
  <ForwardRef(TextComponent)
    as="p"
    text="or"
  />
</div>
`;
